python/lsst/ip/isr/isrMock.py

```diff
@@ -0,0 +1,915 @@
+
+    def makeDefectList(self):
+        r"""Generate a simple single-entry defect list.
+
+        Returns
+        -------
+        defectList : `list` of Defects`
+            Simulated defect list
+        """
+        defectList = []
+        bbox = afwGeom.BoxI(afwGeom.PointI(0, 0),
+                            afwGeom.ExtentI(40, 50))
+        defectList.append(Defect(bbox))
+        return defectList
+
+    def makeCrosstalkCoeff(self):
+        r"""Generate the simulated crosstalk coefficients.
+
+        Returns
+        -------
+        coeffs : `numpy.ndarray`
+            Simulated crosstalk coefficients.
+        """
+
+        return self.crosstalkCoeffs
+
+    def makeTransmissionCurve(self):
+        r"""Generate a simulated flat transmission curve.
+
+        Returns
+        -------
+        transmission : `lsst.afw.image.TransmissionCurve`
+            Simulated transmission curve.
+        """
+
+        return afwImage.TransmissionCurve.makeIdentity()
+
+    def makeImage(self):
+        r"""Generate a simulated ISR image.
+
+        Returns
+        -------
+        exposure : `lsst.afw.image.Exposure` or `dict`
+            Simulated ISR image data.
+
+        Notes
+        -----
+        This method currently constructs a "raw" data image by:
+            * Generating a simulated sky with noise
+            * Adding a single Gaussian "star"
+            * Adding the fringe signal
+            * Multiplying the frame by the simulated flat
+            * Adding dark current (and noise)
+            * Adding a bias offset (and noise)
+            * Adding a overscan gradient parallel to the pixel y-axis
+            * Simulating crosstalk by adding a scaled version of the
+              source amplifier to the target amplifiers.
+
+        The exposure with image data constructed this way is in one of
+        three formats.
+            * A single image, with overscan and prescan regions retained
+            * A single image, with overscan and prescan regions trimmed
+            * A `dict`, containing the amplifer data indexed by the
+              amplifier name.
+
+        The nonlinearity, CTE, and brighter fatter are currently not
+        implemented.
+        """
+        exposure = self.getExposure()
+
+        for idx, amp in enumerate(exposure.getDetector()):
+            bbox = None
+            if self.config.isTrimmed is True:
+                bbox = amp.getBBox()
+            else:
+                bbox = amp.getRawDataBBox()
+
+            ampData = exposure.image[bbox]
+
+            if self.config.doSky is True:
+                self.amplifierAddNoise(ampData, self.config.skyLevel, self.config.skySigma)
+            if self.config.doSource is True:
+                if idx == 0:
+                    self.amplifierAddSource(ampData, self.config.sourceFlux)
+
+            if self.config.doFringe is True:
+                self.amplifierAddFringe(amp, ampData, self.config.fringeScale)
```

Comment:
I think you need to check here that it's is trimmed, because otherwise you'll add fringing to the overscan, and that's not right. Am I thinking correctly here?

Or perhaps, better than asserting that it's trimmed, add logic to only add the fringe to the data region regardless, that would be much better.

Reply:
In case this helps, I wrote this yesterday:

```
def _isExposureTrimmed(exp):
    det = exp.getDetector()
    if exp.getDimensions() == det.getBBox().getDimensions():
        return True
    return False
```

Reply:
I was convinced this was right, looked into it, and discovered we were both wrong.
The `amplifierAddFringe` and `amplifierMultiplyFlat` both need to calculate in the trimmed coordinates, but write the results in the untrimmed coordinates (and, as you mention, not overwrite the overscan).  These methods iterate over the exposure.image[ampBBox].getDimensions() coordinates, which correctly only write to the data region.  However, the conversion method `localCoordToExpCoord` was improperly offsetting things in the untrimmed case (converting from local amp coordinates to untrimmed exposure coordinates, introducing discontinuities across amplifier edges).
This is fixed, and I've confirmed that a fringe made for an untrimmed exposure that is then trimmed by assembleCcd matches a fringe made for a trimmed exposure.

Reply:
OK, sounds good!